syz-cluster: initial code

a-nogikh opened the pull request:
The basic code of a k8s-based cluster that:
* Aggregates new LKML patch series.
* Determines the kernel trees to apply them to.
* Builds the basic and the patched kernel.
* Displays the results on a web dashboard.

This is a very rudimentary version with a lot of TODOs that provides a skeleton for further work.

The project makes use of Argo workflows and Spanner DB.
Bootstrap is used for the web interface.

Overall structure:
* `syz-cluster/dashboard`: a web dashboard listing patch series and their test results.
* `syz-cluster/series-tracker`: polls Lore archives and submits the new patch series to the DB.
* `syz-cluster/controller`: schedules workflows and provides API for them.
* `syz-cluster/kernel-disk`: a cron job that keeps a kernel checkout up to date.
* `syz-cluster/workflow/*`: workflow steps.

For the DB structure see `syz-cluster/pkg/db/migrations/*`.

***

TODO (for this PR):
- [x] Go through the installation section, hopefully get rid of minio use.
- [x] <s>Figure out why `go.mod` bumped to `1.23.1`</s> That is `go: github.com/argoproj/argo-workflows/v3@v3.6.2 requires go >= 1.23.1 (running go 1.22.7)`
- [x] GAE still does not support a Go 1.23 runtime. So should we use an older Argo workflow version? `v3.5.13` still only requires `Go 1.21`.
- [x] `go: module k8s.io/api@v0.32.0 requires go >= 1.23.0; switching to go1.23.4`. `v0.31.4` requires `Go 1.22`.
- [x] Figure out how to run `syz-cluster` tests on GitHub CI (we'll need at least some local Spanner emulator binary).
- [x] Move from minio to https://github.com/fsouza/fake-gcs-server? We might need to use https://argo-workflows.readthedocs.io/en/latest/environment-variables/ to pass `STORAGE_EMULATOR_HOST` to the argo server.

Questions/thoughts out loud:
**1. What to do with the vendor folder**
<s>Our `vendor/` folder is quite big, and this PR adds even more modules on top of that. Do we still want to keep that code in our repostory? Is it possible to only keep the modules needed by other components but not `syz-cluster` (since I do `go mod download` in the Docker containers anyway).</s>
Filed https://github.com/google/syzkaller/issues/5645

**2. Some pkg/ packages are too eager to depend on prog/ and sys/**
It has made some Dockerfiles more complicated that they could have been.
Should we strive to remove these dependencies?

dvyukov commented on pkg/email/lore/parse.go:
```diff
@@ -33,24 +55,137 @@ func Threads(emails []*email.Email) []*Thread {
-var patchSubjectRe = regexp.MustCompile(`\[(?:(?:rfc|resend)\s+)*patch`)
+type PatchSubject struct {
+	Title   string
+	Tags    []string // Sometimes there's e.g. "net" or "next-next" in the subject.
```
These tags are parsed, but lost later. What's the plan here?
Do we want to not test "RFC" patches? Then these should set Corrupted?

a-nogikh replied:
As mentioned in the comment, the developers sometimes explicitly say that the the patch it to be applied to the specific tree. This information is not used in this initial PR, but will be important for the further series triage implementations.

dvyukov commented on pkg/email/lore/parse.go:
```diff
@@ -33,24 +55,137 @@ func Threads(emails []*email.Email) []*Thread {
+			Total:     1,
+			Version:   1,
+		}
+		if patch.Total.IsSet() {
```
`std::optional` has `value_or` method, may be handy here as well.

dvyukov commented on pkg/email/lore/parse.go:
```diff
@@ -20,9 +22,29 @@ type Thread struct {
+	Subject   string
+	MessageID string
+	Version   int
+	Total     int
```
Do we need it here? It's `len(Patches)` unless it's corrupted, right?

dvyukov commented on pkg/email/lore/parse.go:
```diff
@@ -20,9 +22,29 @@ type Thread struct {
 }
 
+type Series struct {
+	Subject   string
```
Does it happen that the subject changes between versions?
If so, do we need to also extract in-reply to message-id, which may be the previous version?

a-nogikh replied:
AFAIK the series are usually not sent in-reply-to the previous ones, so I'd expect the title to be fairly stable.

dvyukov reviewed (commented):
The first patch is not questionable. Let's review/merge it separately.

dvyukov commented on pkg/osutil/osutil.go:
```diff
@@ -279,6 +280,14 @@ func WriteFile(filename string, data []byte) error {
 	return os.WriteFile(filename, data, DefaultFilePerm)
 }
 
+func WriteJSON[T any](filename string, obj T) error {
```
Let's merge this separately as well. There are several uses for it in manager code, maybe more.

a-nogikh replied:
Posted in https://github.com/google/syzkaller/pull/5631.

dvyukov commented on pkg/vcs/git.go:
```diff
@@ -672,3 +567,142 @@ func ParseGitDiff(patch []byte) []string {
+	return err
+}
+
+func (gw GitWrapper) Reset() error {
```
The canonical function comment format understood by godoc/etc:
```
// Reset resets git repo state to a known clean state.
func (gw GitWrapper) Reset() error {
```

dvyukov commented on pkg/vcs/git.go:
```diff
@@ -672,3 +567,142 @@ func ParseGitDiff(patch []byte) []string {
+}
+
+func (gw GitWrapper) Apply(patch []byte) error {
+	// git apply is different from patch in that it should normally handle raw emails with patches.
```
Please convert this to godoc as well:
```
// Apply ...
func (gw GitWrapper) Apply(patch []byte) error {
```

dvyukov commented on pkg/vcs/git.go:
```diff
@@ -672,3 +567,142 @@ func ParseGitDiff(patch []byte) []string {
 	return files
 }
+
+type GitWrapper struct {
```
I would call this type `Git` and rename `git` to something else. The exported type is more important.

dvyukov commented on pkg/vcs/git.go:
```diff
@@ -672,3 +567,142 @@ func ParseGitDiff(patch []byte) []string {
+	Dir      string
+	Precious bool
+	Sandbox  bool
+	Env      []string
```
Can we unexport some of these fields if they are used only within this package?

a-nogikh replied:
I've unexported `Precious`, the others are used outside (at least so far).

dvyukov commented on pkg/vcs/git.go:
```diff
@@ -672,3 +567,142 @@ func ParseGitDiff(patch []byte) []string {
+	ignoreCC map[string]bool
+}
+
+func (gw GitWrapper) Git(args ...string) ([]byte, error) {
```
For exported version it may be more reasonable to call this `Run` or `Exec`.

a-nogikh commented:
FTR: we could add a standalone Spanner emulator binary to our `syz-env` image like described here: https://github.com/GoogleCloudPlatform/cloud-spanner-emulator/blob/master/README.md#via-pre-built-linux-binaries

tarasmadan commented:
> We could add a standalone Spanner emulator binary

Please give more context. Do you need it for the local development or testing?
It should be Ok to use it as a dev environment but I'm hesitating to add it as a CI dependency.
Staging environment may be more reliable way for check your system. And spanner mocks looks like a better (more stable, predictable, faster) alternative for CI testing.

a-nogikh commented:
> > We could add a standalone Spanner emulator binary
> 
> Please give more context. Do you need it for the local development or testing? It should be Ok to use it as a dev environment but I'm hesitating to add it as a CI dependency. Staging environment may be more reliable way for check your system. And spanner mocks looks like a better (more stable, predictable, faster) alternative for CI testing.

For dev environment, it already exists - there's `make run-go-tests` in `syz-cluster` that runs the tests in the k8s environment. I meant running those tests on the GitHub CI as well (ideally by getting rid of `make run-go-tests` as well).

I've just added a [syz-cluster/pkg/db: add support for running under syz-env](https://github.com/google/syzkaller/pull/5620/commits/5c4d2af0f40463e189f3eef8f58909f59940e457) commit that runs the Spanner emulator binary for our Go tests. Seems to be quite simple, actually.

For some other external dependencies (e.g. Argo workflows or the blob storage), there are already mocks in the code, but for the DB I fear that there will just be too much boilerplate code. If we can just run a db emulator for the unit tests, why not do it?

Doing so lets us test not just the higher level logic, but also all SQL commands that are used in the implementations.

tarasmadan commented:
Unit testing allows to test some component in some isolated environment where you can simulate any kind of the external service failure. Mocks give you the required flexibility and are the faster way to emulate external dependencies.
Using the emulator for unit testing we make UT slower and test some interaction... It is not the unit testing by definition... I think it worth discussion...

Emulator may be used to test interaction. In our case we can just use real service instead of the emulator that will always differ from it.

tarasmadan commented:
Let's add the testing and error processing sections to the design doc?

a-nogikh commented:
Update:
* Replaced `minio` (which would be totally irrelevant to our actual setup) by `fake-gcs-server`.
* Tested and fixed the installation instructions (they have now also become simpler).

tarasmadan commented on syz-cluster/pkg/db/migrations/1_initialize.up.sql:
```diff
@@ -0,0 +1,86 @@
+CREATE TABLE Series (
+    ID STRING(36) NOT NULL, -- UUID
+    ExtID STRING(128) NOT NULL, -- For LKML, it's a message ID of the series.
```
Using ExtID as ID we'll remove one column and one index.

a-nogikh replied:
That's an interesting question. Moved it to https://github.com/google/syzkaller/issues/5672.

tarasmadan commented on syz-cluster/db-mgmt/main.go:
```diff
@@ -0,0 +1,83 @@
+		switch os.Args[1] {
+		case "migrate":
+			log.Printf("running schema migrations")
+			err = db.RunMigrations(ctx, uri.Full)
```
We have `gcloud spanner` to run SQL commands. Why is this custom tool better?

a-nogikh replied:
For `migrate`, it's doing more than just run SQL commands -- the https://github.com/golang-migrate/migrate library does rollouts and rollbacks management, which will be very handy once we deploy the system in production.

tarasmadan commented on syz-cluster/db-mgmt/main.go:
```diff
@@ -0,0 +1,83 @@
+		case "migrate":
+			log.Printf("running schema migrations")
+			err = db.RunMigrations(ctx, uri.Full)
+		case "run":
```
Could you please help me to find where `db-mgmt run` is used?

a-nogikh replied:
It's for debugging purposes only (make a query to the DB and see what's inside).

tarasmadan commented on syz-cluster/kernel-disk/cron.yaml:
```diff
@@ -0,0 +1,45 @@
+                requests:
+                  cpu: 4
+                  memory: 8G
+                limits:
```
Is it needed to prevent throttling and OOMs?

tarasmadan replied:
It is better to document then. I used different sources to understand what's going on here.

a-nogikh replied:
I'd say these are very basic K8S/GKE configuration options, we don't really want to duplicate k8s documentation snippets across the code.

Once we start moving to GKE, we'd have to set resource limits for all k8s deployments and jobs.

tarasmadan commented on syz-cluster/dashboard/templates/base.html:
```diff
@@ -0,0 +1,22 @@
+  </head>
+  <body>
+    <nav class="navbar navbar-light bg-light justify-content-start">
+      <a class="navbar-brand px-3" href="#">Syz-Cluster</a>
```
Syz-Cluster looks different. Syzbot, syzkaller, syz-manager etc...
Let's syz-cluster. Syz-cluster because it is the beginning of the line.

tarasmadan commented on syz-cluster/dashboard/handler.go:
```diff
@@ -0,0 +1,162 @@
+// Copyright 2024 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package main
```
The proposal is to minimize the amount of files we're adding to the package "main".
Every new file in "main" package require manual efforts during the github repo downstreaming.

a-nogikh replied:
The `main()` method has to be in the `main` package. If there are several `.go` files in the folder, I think it will be even more confusing if they happen to belong to different packages.

tarasmadan commented on syz-cluster/controller/processor.go:
```diff
@@ -0,0 +1,202 @@
+
+// Do not run more than this number of sessions in parallel.
+// TODO: it'd be different for dev and prod, make it configurable.
+const parallelWorkers = 1
```
Make it the seriesRunner() parameter or move closer to the usage point.

a-nogikh replied:
Made it configurable.

a-nogikh commented on syz-cluster/controller/main.go:
```diff
@@ -0,0 +1,30 @@
+	}()
+	api := NewControllerAPI(env)
+	log.Printf("listening on port 8080")
+	app.Fatalf("listen failed: %v", http.ListenAndServe(":8080", api.Mux()))
```
That's actually what it is -- ListenAndServe is not supposed to exit normally.

tarasmadan commented on syz-cluster/pkg/blob/storage.go:
```diff
@@ -0,0 +1,72 @@
+
+// Storage is not assumed to be used for partciularly large objects (e.g. GB of size),
+// but rather for blobs that risk overwhelming Spanner column size limits.
+type Storage interface {
```
What do you plan to store here?
I have a feeling we can reuse existing components/systems. Gcs? Redis? Spanner?
What do you mean by "overwhelming Spanner column size limits"?

tarasmadan commented on syz-cluster/pkg/workflow/service.go:
```diff
@@ -0,0 +1,58 @@
+)
+
+// MockService serializes callback invocations to simplify test implementations.
+type MockService struct {
```
we have mockery

a-nogikh replied:
Added to https://github.com/google/syzkaller/issues/5672

tarasmadan commented on syz-cluster/pkg/workflow/service.go:
```diff
@@ -0,0 +1,58 @@
+// It's assumed that the workflow will query the necessary data and report its detailed progress itself,
+// so we only need to be able to start it and to check whether its current overall state.
+
+type Service interface {
```
Describe it better please. Workflows look central here.

tarasmadan commented on syz-cluster/pkg/workflow/argo.go:
```diff
@@ -0,0 +1,115 @@
+	switch wf.Status.Phase {
+	case v1alpha1.WorkflowRunning, v1alpha1.WorkflowPending:
+		return StatusRunning, log, nil
+	case v1alpha1.WorkflowSucceeded, v1alpha1.WorkflowFailed:
```
On WorkflowFailed we return StatusFinished. Is it correct?

a-nogikh commented on syz-cluster/pkg/workflow/argo.go:
```diff
@@ -0,0 +1,115 @@
+}
+
+func (w *ArgoService) generateLog(nodes v1alpha1.Nodes) []byte {
+	var list []v1alpha1.NodeStatus
```
It requires Go 1.23

a-nogikh commented on syz-cluster/dashboard/handler.go:
```diff
@@ -0,0 +1,162 @@
+	"github.com/google/syzkaller/syz-cluster/pkg/db"
+)
+
+type DashboardHandler struct {
```
unexported

tarasmadan commented:
lgtm

tarasmadan reviewed (approved):
lgtm

a-nogikh commented on syz-cluster/dashboard/kustomization.yaml:
```diff
@@ -0,0 +1,6 @@
+# Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+resources:
+- deployment.yaml
```
There are not so many of them yet, so let's keep them together for now. Once we move to GKE, we'll likely have to adjust many k8s config files, so we may reconsider the approach.
